src/client/windshaft.js

```diff
@@ -322,45 +328,40 @@ export default class Windshaft {
             body: JSON.stringify(mapConfigAgg),
         };
     }
-    _categorizeString(category) {
-        if (category === undefined) {
-            category = 'null';
-        }
-        if (this._categoryStringToIDMap[category] !== undefined) {
-            return this._categoryStringToIDMap[category];
-        }
-        this._categoryStringToIDMap[category] = this._numCategories;
-        this._numCategories++;
-        return this._categoryStringToIDMap[category];
-    }
 
-    _adaptMetadata(meta) {
+    _adaptMetadata(meta, agg) {
         const { stats, aggregation, dates_as_numbers } = meta;
         const featureCount = stats.hasOwnProperty('featureCount') ? stats.featureCount : stats.estimatedFeatureCount;
         const geomType = adaptGeometryType(stats.geometryType);
-        const columns = Object.keys(stats.columns)
-            .map(name => Object.assign({ name }, stats.columns[name]))
-            .map(col => Object.assign(col, { type: adaptColumnType(col.type) }))
-            .filter(col => ['number', 'date', 'category'].includes(col.type));
-        const categoryIDs = {};
-        columns.forEach(column => {
-            if (column.type === 'category' && column.categories) {
-                column.categories.forEach(category => {
-                    categoryIDs[category.category] = this._categorizeString(category.category);
+
+        const properties = stats.columns;
+        Object.keys(agg.columns).forEach(aggName => {
+            const basename = R.schema.column.getBase(aggName);
+            properties[basename].sourceName = aggName;
+        });
+        Object.values(properties).map(property => {
+            property.type = adaptColumnType(property.type);
+        });
+        const metadata = new Metadata({ properties, featureCount, sample: stats.sample, geomType, isAggregated: aggregation.mvt });
+        Object.keys(properties).forEach(propertyName => {
+            const property = properties[propertyName];
+            if (property.type === 'category' && property.categories) {
+                property.categories.forEach(category => {
+                    metadata.categorizeString(category.category);
```

Comment:
I find it confusing that we have to *categorize* strings here, but the Metadata constructor also does some of this...